internal/cmd/controller/reconciler/cluster_controller.go

```diff
@@ -337,3 +346,42 @@ func (r *ClusterReconciler) mapBundleDeploymentToCluster(ctx context.Context, a
 		},
 	}}
 }
+
+// ensureFinalizer adds a finalizer to a recently created cluster.
+func (r *ClusterReconciler) ensureFinalizer(ctx context.Context, cluster *fleet.Cluster) error {
+	if controllerutil.ContainsFinalizer(cluster, finalize.ClusterFinalizer) {
+
+		return nil
+	}
+
+	controllerutil.AddFinalizer(cluster, finalize.ClusterFinalizer)
+
+	return r.Update(ctx, cluster)
+}
+
+// handleDelete runs cleanup the namespace associated to a Cluster,
+// finally removing the finalizer to unblock the deletion of the object from kubernetes.
+func (r *ClusterReconciler) handleDelete(ctx context.Context, cluster *fleet.Cluster) (ctrl.Result, error) {
+	if !controllerutil.ContainsFinalizer(cluster, finalize.ClusterFinalizer) {
+
+		return ctrl.Result{}, nil
+	}
+
+	if cluster.Status.Namespace != "" {
```

Comment:
nit pick: It might make sense to check if bundle deployments still exist before deleting the namespace to prevent potentially hanging deletions.

Reply:
Yes, but then it is handled by us and we can comment on it, but fine with me leaving it to Kubernetes.